File: leetcode-submissions/0560-subarray-sum-equals-k/solution.java
class Solution {
    public int subarraySum(int[] nums, int k) {
        //hashmap storing the frequency of the prefix

        Map<Integer, Integer> freqs = new HashMap<>();
        int ans = 0;
        int curr = 0;
        freqs.put(0, 1);
        for(int i = 0; i < nums.length; i++) {
            curr += nums[i];
            ans += freqs.getOrDefault(curr - k, 0);
            freqs.put(curr, freqs.getOrDefault(curr, 0) + 1);
        }
        return ans;
    }
}

File: leetcode-submissions/0560-subarray-sum-equals-k/solution.py
class Solution:
    def subarraySum(self, nums: List[int], k: int) -> int:

        hashmap = collections.defaultdict(int)
        hashmap[0] = 1

        cum_sum = 0
        count = 0
        n = len(nums)

        for i in range(n):
            cum_sum += nums[i]

            complement = cum_sum - k
            if complement in hashmap:
                count += hashmap[complement]

            hashmap[cum_sum] += 1

        return count

